Create initial Number utility package

emmadesilva opened the pull request:
## Abstract

Hey everyone! Quick introduction here. I'm Caen, creator of the Number PR of Laravel. In this repo I'm porting that utility so it can be used in all PHP projects. This is my first Friends Of PHP Package, so I'm starting this discussion to ensure that everything matches the style and structure of your other packages.

## Checklist

### Composer package name

I based the package name on the PHP-CS-Fixer package. Is this okay?

```bash
composer require friendsofphp/number
```

### Utility namespace

Is this an appropriate namespace?

```php
use FriendsOfPhp\Number\Number;

Number::format(1234567.89);
```

### Coding styles

Which coding style should we use? Code is currently formatted according to the PER standard: www.php-fig.org/per/coding-style

### Documentation site

How do you want to handle documentation? My go-to for simple packages is to use GitHub Pages to create a static site from the README (or Markdown docs in the repo) using [HydePHP](https://github.com/hydephp/hyde). (For example [hydephp.github.io/action](https://hydephp.github.io/action/) and [git.desilva.se/Speacher](https://git.desilva.se/Speacher) which both just need a short GitHub Actions workflow)

emmadesilva committed:
Merge 'caendesilva/php-number' into 'FriendsOfPHP/number'

emmadesilva commented on composer.json:
```diff
@@ -0,0 +1,28 @@
+{
+    "name": "friendsofphp/number",
```
Is this a good package name? I tried to match the PHP-CS-Fixer.

Wirone replied:
Yes, it's correct vendor name 🙂.

emmadesilva replied:
Thanks!

emmadesilva commented on src/Number.php:
```diff
@@ -0,0 +1,166 @@
+<?php
+
+namespace FOP\Number;
```
Is this an appropriate namespace?

Wirone replied:
I would rather use `FriendsOfPhp\Number`. As far as I see other packages in this org don't use FOP-like top namespace at all, but I believe in this case it's better to add something, because standalone `Number` wouldn't be unique enough.

emmadesilva replied:
Yeah I agree that it becomes kinda clunky to have to import `use VendorName\Number\Number;` as the class is there twice. But I also think it's kinda what we have to do as it's just the way namespaces usually are.

I'm assuming there could become conflicts if we simply used `FriendsOfPhp` as the namespace like this: `use FriendsOfPhp\Number;`, because what if a future package also uses that namespace?

Wirone replied:
Duplicated `Number` does not bother me, because first one is sub-namespace and second one is a class name, it's OK IMHO. I was talking about `use Number\Number;` vs `use FriendsOfPhp\Number\Number;` 🙂. In terms of packages like [Pickle](https://github.com/FriendsOfPHP/pickle) it makes sense to use simple `namespace Pickle;`, but `namespace Number;` is not good enough 🙂.

emmadesilva replied:
Ah got it. Makes total sense! 🙂

emmadesilva committed:
Install PHP-CS-Fixer

emmadesilva committed:
Remove unnecessary example tests

Wirone commented on .gitignore:
```diff
@@ -50,3 +50,9 @@
+.php-cs-fixer.cache
+
+# PhpStorm
+/.idea
```
```suggestion
/.idea/
```

Convention for directories is with trailing `/`

emmadesilva replied:
Thanks!

Wirone commented on README.md:
Examples can be improved:

- assigning to variable is not needed and is redundant (like `$spelledNumber` vs `Number::spell()`)
- instead of commenting the intent, which is clear by the class+method itself, there should be actual output

Like:

```php
echo Number::spell(1234); // one thousand two hundred thirty-four
echo Number::ordinal(42); // 42nd
```

emmadesilva replied:
Resolved in 6456337dd14c1cd3a7e1b008ba2a2a58d69dd389

Wirone commented on composer.json:
```diff
@@ -0,0 +1,29 @@
+{
+    "name": "friendsofphp/number",
+    "description": "Adds a Number utility class",
```
It looks like a commit message 😉. Maybe something more like:

```suggestion
    "description": "Number utility for displaying human-friendly numeric data",
```

emmadesilva replied:
Perfect, thanks!

Wirone commented on composer.json:
```diff
@@ -0,0 +1,29 @@
+        }
+    ],
+    "require": {
+        "php": "^8.0",
```
Since PHP is not SemVer, we should rather limit upper version to the newest available one, which can be tested.

```suggestion
        "php": "^8.0 <8.4",
```

emmadesilva replied:
Cool, didn't know we could do this!

emmadesilva committed:
Change `FOP` namespace to `FriendsOfPhp`

See https://github.com/FriendsOfPHP/number/pull/1#discussion_r1409110993

Wirone commented on tests/Feature/NumberTest.php:
```diff
@@ -0,0 +1,179 @@
+test('ordinal', function () {
+    $this->assertSame('1st', Number::ordinal(1));
+    $this->assertSame('2nd', Number::ordinal(2));
+    $this->assertSame('3rd', Number::ordinal(3));
```
Missing test for `th` suffix:

```suggestion
    $this->assertSame('3rd', Number::ordinal(3));
    $this->assertSame('4th', Number::ordinal(4));
```

Wirone commented on tests/Feature/NumberTest.php:
I am not a Pest user, but I believe the most important part of it is that tests are like english sentences ("it does something"), which is not the case here. Some of the tests are not even grammatically correct. I would recommend:

- using `it()` helper instead of `test()`
- complete descriptions to be a proper sentence that describes the intent, example:

```diff
-test('ordinal', function () {
+it('converts numbers to ordinals', function () {
```

It's not a strict requirement, we can keep `test()` but at least let's make it more like a correct sentence 🙂.

emmadesilva replied:
Good call! I honestly just converted the PHPUnit test I made for Laravel by using multicursor replacements. Will fix it soon (birthday today so working on stuff all over the place 🙂)

emmadesilva replied:
Fixed in 54ddd01fac1667caddc316c1425185e2b8675a97

Wirone commented on tests/Pest.php:
Seems like an unneeded file, there's not any meaningful code here.

emmadesilva replied:
Resolved in 9838f2bd755a340ddf6c906b3eb394e6775b6031

Wirone commented on .github/workflows/tests.yml:
Since there is Fixer installed, we should run it in CI 🙂.

emmadesilva replied:
Will add it!

emmadesilva replied:
Added in f66fa43969b6cb0c8d5ae7a92f2ac89cd17e0372

emmadesilva committed:
Update .gitignore to add trailing slash to directory

Convention for directories is with trailing `/`

Co-authored-by: Greg Korba <wirone@gmail.com>

emmadesilva committed:
Update composer.json package description

Co-authored-by: Greg Korba <wirone@gmail.com>

emmadesilva committed:
Use more helpful Readme examples

From https://github.com/FriendsOfPHP/number/pull/1#discussion_r1409127253

Examples can be improved:

- assigning to variable is not needed and is redundant (like `$spelledNumber` vs `Number::spell()`)
- instead of commenting the intent, which is clear by the class+method itself, there should be actual output

emmadesilva committed:
Delete Pest.php

Not needed as we don't specify anything custom here

See https://github.com/FriendsOfPHP/number/pull/1#discussion_r1409142768

emmadesilva committed:
Delete TestCase.php

Not needed as we don't specify anything custom here

emmadesilva committed:
Create php-cs-fixer.yml

Requested in https://github.com/FriendsOfPHP/number/pull/1#discussion_r1409143812

emmadesilva committed:
Exit action if working tree is clean

emmadesilva committed:
Update Pest tests to use readable descriptions

Uses complete descriptions to be proper sentences that describes the intent

Resolves https://github.com/FriendsOfPHP/number/pull/1#discussion_r1409141154

emmadesilva committed:
Add missing ordinal test for 'th' suffix

Co-authored-by: Greg Korba <wirone@gmail.com>

emmadesilva committed:
Limit upper PHP version

Since PHP is not SemVer, we should rather limit upper version to the newest available one, which can be tested.

Co-Authored-By: Greg Korba <wirone@gmail.com>

emmadesilva committed:
Set minimum PHP version to 8.1

While the code should run fine on 8.0, the libraries we use for testing do not, and I don't think a work around is worth it since 8.0 is no longer supported and 8.1 is the minimum official PHP version.

emmadesilva commented on composer.json:
```diff
@@ -0,0 +1,29 @@
+        }
+    ],
+    "require": {
+        "php": "^8.1 <8.4",
```
**I bumped this to 8.1, is that acceptable?**

Context:

While the code should run fine on 8.0, the libraries we use for testing do not, and I don't think a work around is worth it since 8.0 is no longer supported and 8.1 is the minimum official PHP version.

Wirone replied:
For wider adoption it would be better to support older PHP versions, even 7.4, but it would make everything more complex - CI would have to be configured to run on more PHP versions, dependenciec would have to be downgraded to versions that support older PHP versions, and of course package's code itself wouldn't be able to use newer language features. Hard to tell 🤷‍♂️.

There is always option for downgrading the code using Rector before releasing version, but this is another way of making process more complex 😉.

emmadesilva replied:
Yup, there are always tradeoffs to consider. Maybe a compromise of leaving it as it is now for the initial release, and if we get people actually requesting a backport we can revisit the matter then?

emmadesilva committed:
Add matrix testing for supported PHP versions

emmadesilva commented:
@Wirone Think we could get this published to packagist?

Wirone commented:
It wasn't published 🙈? I don't have rights to the org on Packagist, unfortunately. @dragoonis was coordinating this package, maybe he can help.

emmadesilva commented:
> It wasn't published 🙈? I don't have rights to the org on Packagist, unfortunately. @dragoonis was coordinating this package, maybe he can help.

Nope, realized today it still was not published 🙈🙈 Wasn't sure who to ping

dragoonis commented:
Whoever has Merge rights should merge this to this organisation as it's own repo. 

Who is going to do it?

dragoonis commented:
@Wirone ping the right people.

Wirone commented:
> Whoever has Merge rights should merge this to this organisation as it's own repo. 
> 
> Who is going to do it?

It's already merged, but it needs to be added to Packagist by someone with access to [`friendsofphp`](https://packagist.org/packages/friendsofphp/) vendor. I don't know who's there 😅.
